crates/storage/src/bloom.rs

```diff
@@ -0,0 +1,144 @@
+use std::sync::{Mutex, MutexGuard};
+
+use bloomfilter::Bloom;
+use cached::{Cached, SizedCache};
+use pathfinder_common::{BlockNumber, ReorgCounter};
+use pathfinder_crypto::Felt;
+
+#[derive(Clone)]
+pub(crate) struct BloomFilter(Bloom<Felt>);
+
+impl BloomFilter {
+    const BITMAP_BYTES: u64 = 2048;
+    const ITEMS_COUNT: u32 = 1024;
+    const K_NUM: u32 = 12;
+    const SEED: [u8; 32] = [
+        0xef, 0x51, 0x88, 0x74, 0xef, 0x08, 0x3d, 0xf6, 0x7d, 0x7a, 0x93, 0xb7, 0xb3, 0x13, 0x1f,
+        0x87, 0xd3, 0x26, 0xbd, 0x49, 0xc7, 0x18, 0xcc, 0xe5, 0xd7, 0xe8, 0xa0, 0xdb, 0xea, 0x80,
+        0x67, 0x52,
+    ];
+
+    pub fn new() -> Self {
+        let bloom = Bloom::new_with_seed(
+            Self::BITMAP_BYTES as usize,
+            Self::ITEMS_COUNT as usize,
+            &Self::SEED,
+        );
+        assert_eq!(bloom.number_of_hash_functions(), Self::K_NUM);
+
+        Self(bloom)
+    }
+
+    pub fn from_compressed_bytes(bytes: &[u8]) -> Self {
+        let bytes = zstd::bulk::decompress(bytes, Self::BITMAP_BYTES as usize * 2)
+            .expect("Decompressing Bloom filter");
+        Self::from_bytes(&bytes)
+    }
+
+    pub fn from_bytes(bytes: &[u8]) -> Self {
+        let k1 = u64::from_le_bytes(Self::SEED[0..8].try_into().unwrap());
+        let k2 = u64::from_le_bytes(Self::SEED[8..16].try_into().unwrap());
+        let k3 = u64::from_le_bytes(Self::SEED[16..24].try_into().unwrap());
+        let k4 = u64::from_le_bytes(Self::SEED[24..32].try_into().unwrap());
+        let bloom = Bloom::from_existing(
+            bytes,
+            Self::BITMAP_BYTES * 8,
+            Self::K_NUM,
+            [(k1, k2), (k3, k4)],
+        );
+        Self(bloom)
+    }
+
+    pub fn as_compressed_bytes(&self) -> Vec<u8> {
+        let bytes = self.as_bytes();
+        zstd::bulk::compress(&bytes, 0).expect("Compressing Bloom filter")
+    }
+
+    pub fn as_bytes(&self) -> Vec<u8> {
+        self.0.bitmap()
+    }
+
+    pub fn set(&mut self, key: &Felt) {
+        self.0.set(key);
+    }
+
+    pub fn check(&self, key: &Felt) -> bool {
+        self.0.check(key)
+    }
+}
+
+type CacheKey = (ReorgCounter, BlockNumber);
+pub(crate) struct Cache(Mutex<SizedCache<CacheKey, BloomFilter>>);
+
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum CacheError {
+    #[error("Bloom filter cache lock is poisoned.")]
+    Poisoned,
```

Comment:
Should we use `parking_lot`'s mutex which doesn't have this problem? I know I've wanted to in the past to not have to consider poisoning at all.

Reply:
+1

Reply:
We can also do something like `.unwrap_or_else(|e| e.into_inner())` to get the inner value from the `PoisonError`. A bit ugly as a pattern but maybe better than moving to `parking_lot` just for this?

Reply:
Once poisoned the mutex is broken forever though right? Effectively ending the usefulness of it?

Reply:
No, I don't think that is the case. `PoisonError` contains the same `MutexGuard` you'd get if the mutex were not poisoned.

Reply:
Okay so for both reads and writes you'd always perform the unwrap dance? If so, then that's fine 👍 

We could make it an ergonomics refactor to switch all mutex's to `parking_lot` if we care enough.